### PR TITLE
Respect requested variable order in distinct()

### LINKED ISSUE #6156: distinct() doesn't respect user-specified column order
I'm again experiencing behavior that was reported as https://github.com/tidyverse/dplyr/issues/3195, then closed years ago.  I'll use the same reprex.  Running dplyr 1.0.7.

```
starwars %>% distinct(skin_color, hair_color)

#> # A tibble: 50 x 2
#>    hair_color    skin_color 
#>    <chr>         <chr>      
#>  1 blond         fair       
#>  2 <NA>          gold       
#>  3 <NA>          white, blue
#>  4 none          white      
#>  5 brown         light      
#>  6 brown, grey   light      
#>  7 <NA>          white, red 
#>  8 black         light      
#>  9 auburn, white fair       
#> 10 auburn, grey  fair       
#> # ... with 40 more rows
Created on 2022-01-14 by the reprex package (v2.0.1)
```

### PR DESCRIPTION
**Pull request by hadley**
Fixes #6156

### CODE REVIEW

File: R/distinct.R [hunk 1]
```diff
@@ -107,7 +107,8 @@ distinct_prepare <- function(.data,
   }
 
   # Always include grouping variables preserving input order
```
Comment (DavisVaughan) [inline R/distinct.R:109]:
```suggestion
  # Missing grouping variables are added to the front
```

This seems like a clearer way to talk about what this is doing, right?

I thought about suggesting you do:

```r
new_vars <- unique(c(group_vars, distinct_vars))
```

but after further review I figured out that wasn't right because you want grouping variables that also exist in `distinct_vars` to respect the order they are specified in `distinct_vars`. I think an updated comment would have gotten me there faster.

File: R/distinct.R [hunk 2]
```diff
@@ -107,7 +107,8 @@ distinct_prepare <- function(.data,
...
-  out_vars <- intersect(names(.data), c(distinct_vars, group_vars))
+  new_vars <- c(setdiff(group_vars, distinct_vars), distinct_vars)
+  out_vars <- intersect(new_vars, names(.data))
```
Comment (DavisVaughan) [inline R/distinct.R:111]: I was very curious about what this `intersect()` does. I removed it and this test broke:

```r
test_that("distinct doesn't duplicate columns", {
  df <- tibble(a = 1:3, b = 4:6)

  expect_named(df %>% distinct(a, a), "a")
  expect_named(df %>% group_by(a) %>% distinct(a), "a")
})
```

which I think implies that we were using `intersect()` for its "feature" of removing duplicates

``` r
intersect(c(1, 1), 1)
#> [1] 1
```

So probably `unique(new_vars)` is enough and more correct there?

If you think about it, both `distinct_vars` and `group_vars` should always contain cols in `.data`. We explicitly check that this is the case for `distinct_vars` (thats the `missing_vars` bit). Maybe we should change to `unique(new_vars)` and then add an explicit check that `group_vars` are in `names(.data)` (since it is user input) and error if that isn't true.

File: tests/testthat/test-distinct.R [hunk 3]
```diff
@@ -102,9 +102,15 @@ test_that("distinct handles 0 columns edge case (#2954)", {
   expect_equal(nrow(distinct(tibble())), 0L)
 })
 
-test_that("distinct preserves order of the input variables (#3195)",{
+test_that("distinct respects order of the specified variables (#3195)",{
```
Comment (DavisVaughan) [inline tests/testthat/test-distinct.R:105]:
```suggestion
test_that("distinct respects order of the specified variables (#3195, #6156)",{
```

Maybe worth referencing both issues

**Review by DavisVaughan — APPROVED** [review]

File: R/distinct.R (hunk 2 above)
Comment (hadley) [reply R/distinct.R:111]: I don't think we need to check `group_vars()` because it's always computed from the input data. But I agree that `unique()` is clearer than `intersect()`

Comment (DavisVaughan) [reply R/distinct.R]: I only mentioned that we check it because `distinct_prepare()` is exported (for whatever reason) and `group_vars` is an argument, so who knows what someone might pass in. Internally I don't think we will ever have an issue.

It is also fine to not check it and just assume that only advanced developers will ever use this, and they should know what they are doing
